lms/services/grouping.py

```diff
@@ -26,74 +26,48 @@ def upsert(self, grouping):
+        tool_consumer_instance_guid, lms_id, parent: Grouping, type_: Grouping.Type
     ):
-        """
-        Upsert a Grouping for a course section.
-
-        :param tool_consumer_instance_guid: Tool consumer GUID
-        :param context_id: Course id the section is a part of
-        :param section_id: A section id for a section group
-        :param section_name: The name of the section
-        """
-
-        section_authority_provided_id = hashed_id(
-            tool_consumer_instance_guid, context_id, section_id
-        )
-
-        course = self._course_service.get(
-            self._course_service.generate_authority_provided_id(
-                tool_consumer_instance_guid, context_id
+        if type_ == Grouping.Type.CANVAS_SECTION:
+            return hashed_id(tool_consumer_instance_guid, parent.lms_id, lms_id)
+        if type_ == Grouping.Type.COURSE:
+            return CourseService.generate_authority_provided_id(
+                tool_consumer_instance_guid, lms_id
             )
-        )
 
-        return self.upsert(
-            CanvasSection(
-                application_instance=self._application_instance,
-                authority_provided_id=section_authority_provided_id,
-                lms_id=section_id,
-                lms_name=section_name,
-                parent_id=course.id,
-            )
-        )
+        return hashed_id(tool_consumer_instance_guid, parent.lms_id, type_, lms_id)
 
-    def upsert_canvas_group(  # pylint: disable=too-many-arguments
+    def upsert_with_parent(  # pylint: disable=too-many-arguments
         self,
         tool_consumer_instance_guid,
-        context_id,
-        group_id,
-        group_name,
-        group_set_id,
+        lms_id,
+        lms_name,
+        parent: Grouping,
+        type_: Grouping.Type,
+        extra=None,
     ):
         """
-        Upserst a Grouping for a canvas group.
+        Upsert a Grouping generating the authority_provided_id based on its parent.
 
         :param tool_consumer_instance_guid: Tool consumer GUID
-        :param context_id: Course id the group is a part of
-        :param group_id: Canvas group id
-        :param group_name: The name of the group
-        :param group_set_id: Id of the canvas group set this group belongs to
+        :param lms_id: ID of this grouping on the LMS
+        :param lms_name: Name of the grouping on the LMS
+        :param parent: Parent of grouping
+        :param type_: Type of the grouping
+        :param extra: Any extra information to store linked to this grouping
         """
-
-        group_authority_provided_id = hashed_id(
-            tool_consumer_instance_guid, context_id, "canvas_group", group_id
-        )
-
-        course_authority_provided_id = hashed_id(
-            tool_consumer_instance_guid, context_id
-        )
-
-        course = self._course_service.get(course_authority_provided_id)
-
         return self.upsert(
             CanvasGroup(
```

Comment:
It looks like `upsert_with_parent()` _always_ creates a `CanvasGroup` object?

Reply:
> It looks like `upsert_with_parent()` _always_ creates a `CanvasGroup` object?

Oh this is very wrong, I'll check why a bunch of tests haven't exploded.